Source/ACE/Network/GameAction/Actions/GameActionChangeCombatMode.cs

```diff
@@ -9,7 +10,7 @@ public static class GameActionChangeCombatMode
         public static void Handle(ClientMessage message, Session session)
         {
             var newCombatMode = message.Payload.ReadUInt32();
-            session.Player.SetCombatMode((CombatMode)newCombatMode);
+            session.Player.SetCombatMode((CombatMode)newCombatMode, session.Player);
```

Comment:
No, your close to correct -- just this one part needs to be fixed up.

GameAction messages come in from the network (and are therefore run in network threads).  In order for the network threads to interact with game state safely, we need to add some synchronization.

By convention I've done this in Player.cs by creating functions named HandleAction<thing>, which use an ActionChain to move the work onto the player.  That should be an acceptable solution here to solve your concurrency issues (if you move the logic in the ChangeCombatMode to the appropriate place it should take care of your issues there too).